### PR TITLE
use skopeo to fetch blobs

### PR DESCRIPTION
**Pull request by runcom**
@philips @rhatdan @mtrmac

@philips can you ping all interested ppl?

Pls refer to just the first commit as the second is just vendoring skopeo itself. 

This is a basic POC (which just works) about using skopeo libraries in docker2aci for instance (as well as k8s and such)

This POC just gets rid of the layer call, eventually docker2aci should not bother anymore about pulling image and you can see how clean it looks like.

it's also a breeze to remove all those loc which otherwise needs to be maintained - shifting responsability to skopeo for fetching is good also.

```
19:30:12 amurdaca at localhost in github.com/appc/docker2aci ‹master*› ./bin/docker2aci docker://fedora
INFO[0002] Downloading library/fedora/blobs/sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4 
INFO[0002] Downloading library/fedora/blobs/sha256:8b53fe19c255ee721e5b2441b8469fc01a6324e8564de89bf8485387b34f95e6 
Downloading sha256:8b53fe19c25 [===================================] 76.7 MB / ?
Downloading sha256:a3ed95caeb0 [===================================]    32 B / ?

Generated ACI(s):
library-fedora-latest.aci
```

Pull requests like #176 can be entirely skipped here and skopeo can just take care of it (v2s2 docker manifest is supported in skopeo for instance and adding OCI is plain trivial in skopeo - you just ask for an OCI manifest)

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

### CODE REVIEW

File: lib/backend/repository/repository2.go [hunk 1]
```diff
@@ -24,15 +24,15 @@ import (
 	"os"
 	"path"
 	"regexp"
-	"strconv"
 	"strings"
 	"time"
 
+	"github.com/appc/docker2aci/gopath/src/github.com/coreos/ioprogress"
```
Comment (runcom) [inline lib/backend/repository/repository2.go]: not sure why this popped out

File: lib/internal/backend/repository/repository2.go [hunk 2]
```diff
@@ -294,49 +294,17 @@ func getLayerIndex(layerID string, manifest v2Manifest) (int, error) {
...
-		if location != "" {
-			req, err = http.NewRequest("GET", location, nil)
-			if err != nil {
-				return nil, nil, err
-			}
-			res, err = rb.makeRequest(req, dockerURL.ImageName)
-			if err != nil {
-				return nil, nil, err
-			}
-			defer res.Body.Close()
```
Comment (runcom) [inline lib/internal/backend/repository/repository2.go]: this is probably a bug in `master` as I don't see how can you return a closed response body at the end :confused:

Comment (dgonyeo) [inline lib/internal/backend/repository/repository2.go]: I think you might be right.

**Comment by dgonyeo** [comment]
You'll need to vendor skopeo with godeps for the semaphore build to succeed.

**Comment by runcom** [comment]
> You'll need to vendor skopeo with godeps for the semaphore build to succeed.

I'll fix it asap

**Comment by runcom** [comment]
@dgonyeo I'm getting lots of troubles with Godep right now as it's removing stuff which I don't think they should - could you try to vendor skopeo/docker in this PR?

**Comment by dgonyeo** [comment]
Sure, godeps can be a pain.

**Comment by dgonyeo** [comment]
@runcom add this commit to your PR: https://github.com/dgonyeo/docker2aci/commit/952dff7b273a0d7896d0473555832fbe2d85e193

**Comment by runcom** [comment]
Why it says "not master"? You can just vendor skopeo master

**Comment by dgonyeo** [comment]
It was because https://github.com/projectatomic/skopeo/pull/104 wasn't merged, but it appears it is now. Feel free to remove that from the commit message.

**Comment by runcom** [comment]
Done, this should build cleanly now

**Comment by runcom** [comment]
it does

**Comment by yifan-gu** [comment]
cc @euank @tmrts

**Comment by runcom** [comment]
added two more commits, the first takes care of getting the manifest via skopeo, second remove an O(N) routine in favor of an hashmap

**Comment by philips** [comment]
We need to sort out the license of skopeo first. Just noticed it is gplv2.

**Comment by euank** [comment]
It looks like skopeo is now Apache 2.0, though there also should be a release tagged and a re-vendoring here as well.

**Comment by runcom** [comment]
I think it's better to now start using libraries from containers/image - and for the record we'll soon add a sub-library to make appc->docker conversion in there. I'll rework this PR

**Comment by runcom** [comment]
I don't think this is needed anymore - work is being done in github.com/containers/image and we're working on pulling docker -> aci format with @sgotti 

I've also noticed https://github.com/appc/docker2aci/pull/176 has been merged adding lots of duplication with the various manifests format which containers/image does not have so I'll close this in favor of working together toward using that library

**Pull request closed without merging**